File: Windows/BarkFluff.Client.Core/Markdown/MarkdownDocument.cs
namespace BarkFluff.Client.Core.Markdown;

/// <summary>
/// Разобранное сообщение. Слой без UI: разбор живёт здесь, потому что тем же парсером
/// пользуется <see cref="MarkdownText.Strip"/> для plain-text превью, а превью строятся
/// во вьюмоделях.
/// </summary>
public sealed record MarkdownDocument(IReadOnlyList<MarkdownBlock> Blocks)
{
    /// <summary>
    /// Сообщение без разметки: одна строка обычного текста без стилей и ссылок.
    /// Такие сообщения составляют основную массу ленты, и рендер отдаёт их простым
    /// <c>TextBlock</c> вместо дерева блоков.
    /// </summary>
    public bool TryGetPlainText(out string text)
    {
        text = string.Empty;
        if (Blocks is not [MarkdownTextGroup { Alignment: MarkdownAlignment.Start } group]
            || group.Lines is not [{ Kind: MarkdownLineKind.Paragraph } line])
        {
            return false;
        }

        if (line.Inlines.Any(inline => inline.Style is not MarkdownInlineStyle.None || inline.Link is not null))
        {
            return false;
        }

        text = string.Concat(line.Inlines.Select(inline => inline.Text));
        return true;
    }
}

public abstract record MarkdownBlock;

/// <summary>Подряд идущие строки обычного текста, заголовков, списков и разделителей.</summary>
public sealed record MarkdownTextGroup(IReadOnlyList<MarkdownLine> Lines, MarkdownAlignment Alignment) : MarkdownBlock;

/// <summary>Ограждённый блок кода. Info-string после <c>```</c> отбрасывается: подсветки нет ни в одном клиенте.</summary>
public sealed record MarkdownCodeBlock(string Code) : MarkdownBlock;

/// <summary>Подряд идущие строки цитаты.</summary>
public sealed record MarkdownQuote(IReadOnlyList<MarkdownLine> Lines) : MarkdownBlock;

public sealed record MarkdownTable(
    IReadOnlyList<MarkdownTableCell> Headers,
    IReadOnlyList<MarkdownAlignment> Alignments,
    IReadOnlyList<MarkdownTableRow> Rows) : MarkdownBlock;

public sealed record MarkdownTableRow(IReadOnlyList<MarkdownTableCell> Cells);

public sealed record MarkdownTableCell(IReadOnlyList<MarkdownInline> Inlines);

/// <summary>Изображение из HTML-подмножества. <c>Url</c> равен null, если адрес не прошёл allowlist.</summary>
public sealed record MarkdownImage(
    string? Url,
    string Alt,
    int? Width,
    int? Height,
    MarkdownAlignment Alignment,
    string? LinkUrl) : MarkdownBlock;

/// <summary>Строка внутри <see cref="MarkdownTextGroup"/> или <see cref="MarkdownQuote"/>.</summary>
public sealed record MarkdownLine(
    MarkdownLineKind Kind,
    IReadOnlyList<MarkdownInline> Inlines,
    int HeadingLevel = 0,
    string OrderedMarker = "");

/// <summary>Отрезок текста с наложенными стилями. Стили — флаги, как span'ы в Android.</summary>
public sealed record MarkdownInline(string Text, MarkdownInlineStyle Style, string? Link);

public enum MarkdownAlignment
{
    Start,
    Center,
    End
}

public enum MarkdownLineKind
{
    Paragraph,
    Heading,
    Bullet,
    Ordered,
    Rule
}

[Flags]
public enum MarkdownInlineStyle
{
    None = 0,
    Bold = 1,
    Italic = 2,
    Strikethrough = 4,
    Code = 8,
    Small = 16
}

File: Mac/Barkfluff/Packages/BFCore/Sources/BFCore/Markdown/MarkdownDocument.swift
import Foundation

/// Parsed message content shared by the Apple clients.
public struct MarkdownDocument: Hashable, Sendable {
    public let blocks: [MarkdownBlock]

    public init(blocks: [MarkdownBlock]) {
        self.blocks = blocks
    }

    /// Returns the original content when the document contains no formatting.
    /// This lets message lists render the common plain-text case without walking
    /// the complete block tree.
    public var plainText: String? {
        guard !blocks.isEmpty else { return "" }
        guard blocks.count == 1,
              case let .text(group) = blocks[0],
              group.alignment == .start,
              group.lines.allSatisfy({ line in
                  line.kind == .paragraph
                      && line.inlines.allSatisfy { inline in
                          inline.style == .none && inline.linkURL == nil
                      }
              }) else {
            return nil
        }

        return group.lines
            .map { $0.inlines.map(\.text).joined() }
            .joined(separator: "\n")
    }
}

public enum MarkdownBlock: Hashable, Sendable {
    case text(MarkdownTextGroup)
    case code(String)
    case quote([MarkdownLine])
    case table(MarkdownTable)
    case image(MarkdownImage)
}

public struct MarkdownTextGroup: Hashable, Sendable {
    public let lines: [MarkdownLine]
    public let alignment: MarkdownAlignment

    public init(lines: [MarkdownLine], alignment: MarkdownAlignment = .start) {
        self.lines = lines
        self.alignment = alignment
    }
}

public struct MarkdownLine: Hashable, Sendable {
    public let kind: MarkdownLineKind
    public let inlines: [MarkdownInline]
    public let headingLevel: Int
    public let orderedMarker: String

    public init(
        kind: MarkdownLineKind,
        inlines: [MarkdownInline],
        headingLevel: Int = 0,
        orderedMarker: String = ""
    ) {
        self.kind = kind
        self.inlines = inlines
        self.headingLevel = headingLevel
        self.orderedMarker = orderedMarker
    }
}

public enum MarkdownLineKind: Hashable, Sendable {
    case paragraph
    case heading
    case bullet
    case ordered
    case rule
}

public struct MarkdownInline: Hashable, Sendable {
    public let text: String
    public let style: MarkdownInlineStyle
    public let linkURL: String?

    public init(
        text: String,
        style: MarkdownInlineStyle = .none,
        linkURL: String? = nil
    ) {
        self.text = text
        self.style = style
        self.linkURL = linkURL
    }
}

public struct MarkdownInlineStyle: OptionSet, Hashable, Sendable {
    public let rawValue: UInt8

    public init(rawValue: UInt8) {
        self.rawValue = rawValue
    }

    public static let none = MarkdownInlineStyle([])
    public static let bold = MarkdownInlineStyle(rawValue: 1 << 0)
    public static let italic = MarkdownInlineStyle(rawValue: 1 << 1)
    public static let strikethrough = MarkdownInlineStyle(rawValue: 1 << 2)
    public static let code = MarkdownInlineStyle(rawValue: 1 << 3)
    public static let small = MarkdownInlineStyle(rawValue: 1 << 4)
}

public enum MarkdownAlignment: Hashable, Sendable {
    case start
    case center
    case end
}

public struct MarkdownTable: Hashable, Sendable {
    public let headers: [MarkdownTableCell]
    public let alignments: [MarkdownAlignment]
    public let rows: [MarkdownTableRow]

    public init(
        headers: [MarkdownTableCell],
        alignments: [MarkdownAlignment],
        rows: [MarkdownTableRow]
    ) {
        self.headers = headers
        self.alignments = alignments
        self.rows = rows
    }
}

public struct MarkdownTableRow: Hashable, Sendable {
    public let cells: [MarkdownTableCell]

    public init(cells: [MarkdownTableCell]) {
        self.cells = cells
    }
}

public struct MarkdownTableCell: Hashable, Sendable {
    public let inlines: [MarkdownInline]

    public init(inlines: [MarkdownInline]) {
        self.inlines = inlines
    }
}

/// An image parsed from the safe HTML subset.
public struct MarkdownImage: Hashable, Sendable {
    public let url: String?
    public let alt: String
    public let width: Int?
    public let height: Int?
    public let alignment: MarkdownAlignment
    public let linkURL: String?

    public init(
        url: String?,
        alt: String,
        width: Int? = nil,
        height: Int? = nil,
        alignment: MarkdownAlignment = .start,
        linkURL: String? = nil
    ) {
        self.url = url
        self.alt = alt
        self.width = width
        self.height = height
        self.alignment = alignment
        self.linkURL = linkURL
    }
}
